Graph types as direct PR (#1406)
neo4j/docs-cypher: modules/ROOT/pages/schema/constraints/list-constraints.adoc

All constraints created when xref:schema/graph-types/set-graph-types.adoc[setting a graph type] will be returned by the `SHOW CONSTRAINTS` command.
Similarly, all constraints created using the xref:schema/constraints/create-constraints.adoc[`CREATE CONSTRAINT`] syntax will be returned by the `SHOW CURRENT GRAPH TYPE` command.
For more information, see the following sections of the xref:schema/graph-types/list-graph-types[Show graph types] page:

* xref:schema/graph-types/list-graph-types.adoc#constraints-in-show-current-graph-type[Show graph types ->  Constraints in `SHOW CURRENT GRAPH TYPE`]
* xref:schema/graph-types/list-graph-types.adoc#graph-type-elements-in-show-constraints[Show graph types -> Graph type elements in `SHOW CONSTRAINTS`]
====

One of the output columns from `SHOW CONSTRAINTS` is the name of the constraint.
This can be used to drop the constraint with the xref::schema/constraints/drop-constraints.adoc[`DROP CONSTRAINT` command].

[NOTE]
Listing constraints requires the link:{neo4j-docs-base-uri}/operations-manual/current/authentication-authorization/database-administration/#access-control-database-administration-constraints[`SHOW CONSTRAINTS` privilege].

////

> The below constraints are created on schema/constraints/create-constraints.adoc

Reviewer: I think you need to add the vector property type constraints to this list, don't you? (copying them over from the examples in the create file)

Reviewer: I added a suggestion for where they should go to keep the order from the create file

Author: thanks a lot

[source, cypher, role=test-setup]
----
CREATE CONSTRAINT book_isbn
FOR (book:Book) REQUIRE book.isbn IS UNIQUE;
CREATE CONSTRAINT sequels
FOR ()-[sequel:SEQUEL_OF]-() REQUIRE sequel.order IS UNIQUE;
CREATE CONSTRAINT book_title_year
FOR (book:Book) REQUIRE (book.title, book.publicationYear) IS UNIQUE;
CREATE CONSTRAINT prequels
FOR ()-[prequel:PREQUEL_OF]-() REQUIRE (prequel.order, prequel.author) IS UNIQUE;
CREATE CONSTRAINT author_name
FOR (author:Author) REQUIRE author.name IS NOT NULL;
CREATE CONSTRAINT wrote_year

> FOR ()-[wrote:WROTE]-() REQUIRE wrote.year IS NOT NULL;
> CREATE CONSTRAINT movie_title
> FOR (movie:Movie) REQUIRE movie.title IS :: STRING;
> CREATE CONSTRAINT part_of
> FOR ()-[part:PART_OF]-() REQUIRE part.order IS :: INTEGER;
> CREATE CONSTRAINT movie_tagline

Reviewer: This is the spot where they would have been added in the create file:
```suggestion
FOR ()-[part:PART_OF]-() REQUIRE part.order IS :: INTEGER;
CREATE CONSTRAINT node_vector_constraint
FOR (n:Movie) REQUIRE n.embedding IS :: VECTOR<INT32>(42);
CREATE CONSTRAINT rel_vector_constraint
FOR ()-[r:CONTAINS]->() REQUIRE r.embedding IS :: VECTOR<FLOAT32>(1536);
CREATE CONSTRAINT movie_tagline
```

CREATE CONSTRAINT part_of_tags
FOR ()-[part:PART_OF]-() REQUIRE part.tags IS :: STRING | LIST<STRING NOT NULL>;
CREATE CONSTRAINT director_imdbId
FOR (director:Director) REQUIRE (director.imdbId) IS NODE KEY;
CREATE CONSTRAINT ownershipId
FOR ()-[owns:OWNS]-() REQUIRE owns.ownershipId IS RELATIONSHIP KEY;
CREATE CONSTRAINT actor_fullname
FOR (actor:Actor) REQUIRE (actor.firstname, actor.surname) IS NODE KEY;
CREATE CONSTRAINT knows_since_how
FOR ()-[knows:KNOWS]-() REQUIRE (knows.since, knows.how) IS RELATIONSHIP KEY;
CREATE CONSTRAINT node_uniqueness_param
FOR (book:Book) REQUIRE book.prop1 IS UNIQUE;
CREATE CONSTRAINT rel_exist_param
FOR ()-[wrote:WROTE]-() REQUIRE wrote.published IS NOT NULL
----
////

.List all constraints with default output columns
======

.Query
// tag::schema_constraints_show[]
[source, cypher, test-exclude-cols=id]
----
SHOW CONSTRAINTS
----
// end::schema_constraints_show[]

.Result
[source, queryresult]
----
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+

> | id | name                     | type                               | entityType     | labelsOrTypes  | properties                   | ownedIndex              | propertyType                     |

Reviewer: Missing the `enforcedLabel` column

Author: thank you for spotting this - for whatever reason, i used the old tables here. fixing...

> .Result
> [source, queryresult]
> ----
> +--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
> | id | name                     | type                               | entityType     | labelsOrTypes  | properties                   | ownedIndex              | propertyType                     |
> +--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
> | 23 | "actor_fullname"         | "NODE_KEY"                         | "NODE"         | ["Actor"]      | ["firstname", "surname"]     | "actor_fullname"        | NULL                             |
> | 12 | "author_name"            | "NODE_PROPERTY_EXISTENCE"          | "NODE"         | ["Author"]     | ["name"]                     | NULL                    | NULL                             |
> | 5  | "book_isbn"              | "NODE_PROPERTY_UNIQUENESS"         | "NODE"         | ["Book"]       | ["isbn"]                     | "book_isbn"             | NULL                             |
> | 9  | "book_title_year"        | "NODE_PROPERTY_UNIQUENESS"         | "NODE"         | ["Book"]       | ["title", "publicationYear"] | "book_title_year"       | NULL                             |
> | 19 | "director_imdbId"        | "NODE_KEY"                         | "NODE"         | ["Director"]   | ["imdbId"]                   | "director_imdbId"       | NULL                             |
> | 25 | "knows_since_how"        | "RELATIONSHIP_KEY"                 | "RELATIONSHIP" | ["KNOWS"]      | ["since", "how"]             | "knows_since_how"       | NULL                             |
> | 16 | "movie_tagline"          | "NODE_PROPERTY_TYPE"               | "NODE"         | ["Movie"]      | ["tagline"]                  | NULL                    | "STRING | LIST<STRING NOT NULL>" |
> | 14 | "movie_title"            | "NODE_PROPERTY_TYPE"               | "NODE"         | ["Movie"]      | ["title"]                    | NULL                    | "STRING"                         |
> | 27 | "node_uniqueness_param"  | "NODE_PROPERTY_UNIQUENESS"         | "NODE"         | ["Book"]       | ["prop1"]                    | "node_uniqueness_param" | NULL                             |
> | 2  | "node_vector_constraint" | "NODE_PROPERTY_TYPE"               | "NODE"         | ["Movie"]      | ["embedding"]                | NULL                    | "VECTOR<INTEGER32 NOT NULL>(42)" |
> | 21 | "ownershipId"            | "RELATIONSHIP_KEY"                 | "RELATIONSHIP" | ["OWNS"]       | ["ownershipId"]              | "ownershipId"           | NULL                             |
> | 15 | "part_of"                | "RELATIONSHIP_PROPERTY_TYPE"       | "RELATIONSHIP" | ["PART_OF"]    | ["order"]                    | NULL                    | "INTEGER"                        |
> | 17 | "part_of_tags"           | "RELATIONSHIP_PROPERTY_TYPE"       | "RELATIONSHIP" | ["PART_OF"]    | ["tags"]                     | NULL                    | "STRING | LIST<STRING NOT NULL>" |
> | 11 | "prequels"               | "RELATIONSHIP_PROPERTY_UNIQUENESS" | "RELATIONSHIP" | ["PREQUEL_OF"] | ["order", "author"]          | "prequels"              | NULL                             |
> | 28 | "rel_exist_param"        | "RELATIONSHIP_PROPERTY_EXISTENCE"  | "RELATIONSHIP" | ["WROTE"]      | ["published"]                | NULL                    | NULL                             |
> | 3  | "rel_vector_constraint"  | "RELATIONSHIP_PROPERTY_TYPE"       | "RELATIONSHIP" | ["CONTAINS"]   | ["embedding"]                | NULL                    | "VECTOR<FLOAT32 NOT NULL>(1536)" |
> | 7  | "sequels"                | "RELATIONSHIP_PROPERTY_UNIQUENESS" | "RELATIONSHIP" | ["SEQUEL_OF"]  | ["order"]                    | "sequels"               | NULL                             |
> | 13 | "wrote_year"             | "RELATIONSHIP_PROPERTY_EXISTENCE"  | "RELATIONSHIP" | ["WROTE"]      | ["year"]                     | NULL                    | NULL                             |
> +--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
> ----

Author: ```suggestion
.Result
[source, queryresult]
----
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| id | name                    | type                               | entityType     | labelsOrTypes  | properties                   | enforcedLabel | ownedIndex              | propertyType                     |
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| 28 | "actor_fullname"        | "NODE_KEY"                         | "NODE"         | ["Actor"]      | ["firstname", "surname"]     | NULL          | "actor_fullname"        | NULL                             |
| 27 | "author_name"           | "NODE_PROPERTY_EXISTENCE"          | "NODE"         | ["Author"]     | ["name"]                     | NULL          | NULL                    | NULL                             |
| 34 | "book_isbn"             | "NODE_PROPERTY_UNIQUENESS"         | "NODE"         | ["Book"]       | ["isbn"]                     | NULL          | "book_isbn"             | NULL                             |
| 23 | "book_title_year"       | "NODE_PROPERTY_UNIQUENESS"         | "NODE"         | ["Book"]       | ["title", "publicationYear"] | NULL          | "book_title_year"       | NULL                             |
| 36 | "director_imdbId"       | "NODE_KEY"                         | "NODE"         | ["Director"]   | ["imdbId"]                   | NULL          | "director_imdbId"       | NULL                             |
| 26 | "knows_since_how"       | "RELATIONSHIP_KEY"                 | "RELATIONSHIP" | ["KNOWS"]      | ["since", "how"]             | NULL          | "knows_since_how"       | NULL                             |
| 12 | "movie_tagline"         | "NODE_PROPERTY_TYPE"               | "NODE"         | ["Movie"]      | ["tagline"]                  | NULL          | NULL                    | "STRING | LIST<STRING NOT NULL>" |
| 33 | "movie_title"           | "NODE_PROPERTY_TYPE"               | "NODE"         | ["Movie"]      | ["title"]                    | NULL          | NULL                    | "STRING"                         |
| 17 | "node_uniqueness_param" | "NODE_PROPERTY_UNIQUENESS"         | "NODE"         | ["Book"]       | ["prop1"]                    | NULL          | "node_uniqueness_param" | NULL                             |
| 38 | "ownershipId"           | "RELATIONSHIP_KEY"                 | "RELATIONSHIP" | ["OWNS"]       | ["ownershipId"]              | NULL          | "ownershipId"           | NULL                             |
| 41 | "part_of"               | "RELATIONSHIP_PROPERTY_TYPE"       | "RELATIONSHIP" | ["PART_OF"]    | ["order"]                    | NULL          | NULL                    | "INTEGER"                        |
| 39 | "part_of_tags"          | "RELATIONSHIP_PROPERTY_TYPE"       | "RELATIONSHIP" | ["PART_OF"]    | ["tags"]                     | NULL          | NULL                    | "STRING | LIST<STRING NOT NULL>" |
| 24 | "prequels"              | "RELATIONSHIP_PROPERTY_UNIQUENESS" | "RELATIONSHIP" | ["PREQUEL_OF"] | ["order", "author"]          | NULL          | "prequels"              | NULL                             |
| 30 | "rel_exist_param"       | "RELATIONSHIP_PROPERTY_EXISTENCE"  | "RELATIONSHIP" | ["WROTE"]      | ["published"]                | NULL          | NULL                    | NULL                             |
| 19 | "sequels"               | "RELATIONSHIP_PROPERTY_UNIQUENESS" | "RELATIONSHIP" | ["SEQUEL_OF"]  | ["order"]                    | NULL          | "sequels"               | NULL                             |
| 14 | "wrote_year"            | "RELATIONSHIP_PROPERTY_EXISTENCE"  | "RELATIONSHIP" | ["WROTE"]      | ["year"]                     | NULL          | NULL                    | NULL                             |
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
----
```

======

.List all constraints with full details
======

To return the full details of the constraints on a database, use `SHOW CONSTRAINTS YIELD *`

.List all constraints with `YIELD *`
// tag::schema_constraints_show_full[]
[source, cypher, test-exclude-cols=id]
----
SHOW CONSTRAINTS YIELD *
----
// end::schema_constraints_show_full[]

.Result
[source, queryresult]
----
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+

> | id | name                     | type                               | entityType     | labelsOrTypes  | properties                   | ownedIndex              | propertyType                     | options           | createStatement                                                                                                                   |

Reviewer: missing the `enforcedLabel` and `classification` columns

| 12 | "movie_tagline"         | "NODE_PROPERTY_TYPE"               | "NODE"         | ["Movie"]      | ["tagline"]                  | NULL          | "independent"  | NULL                    | "STRING | LIST<STRING NOT NULL>" | NULL              | "CREATE CONSTRAINT `movie_tagline` FOR (n:`Movie`) REQUIRE (n.`tagline`) IS :: STRING | LIST<STRING NOT NULL>"     |
| 33 | "movie_title"           | "NODE_PROPERTY_TYPE"               | "NODE"         | ["Movie"]      | ["title"]                    | NULL          | "independent"  | NULL                    | "STRING"                         | NULL              | "CREATE CONSTRAINT `movie_title` FOR (n:`Movie`) REQUIRE (n.`title`) IS :: STRING"                                 |
| 17 | "node_uniqueness_param" | "NODE_PROPERTY_UNIQUENESS"         | "NODE"         | ["Book"]       | ["prop1"]                    | NULL          | "undesignated" | "node_uniqueness_param" | NULL                             | {indexConfig: {}} | "CREATE CONSTRAINT `node_uniqueness_param` FOR (n:`Book`) REQUIRE (n.`prop1`) IS UNIQUE"                           |

> | 2  | "node_vector_constraint" | "NODE_PROPERTY_TYPE"               | "NODE"         | ["Movie"]      | ["embedding"]                | NULL         | ""             | NULL                    | "VECTOR<INTEGER32 NOT NULL>(42)" | NULL              | "CREATE CONSTRAINT `node_vector_constraint` FOR (n:`Movie`) REQUIRE (n.`embedding`) IS :: VECTOR<INTEGER32 NOT NULL>(42)"         |

Author: what's the value for "classification" here?

Reviewer: That depends on how they get created but assuming they are created using the `CREATE CONSTRAINT` command, it would be `"independent"`
```suggestion
| 2  | "node_vector_constraint" | "NODE_PROPERTY_TYPE"               | "NODE"         | ["Movie"]      | ["embedding"]                | NULL         | "independent"             | NULL                    | "VECTOR<INTEGER32 NOT NULL>(42)" | NULL              | "CREATE CONSTRAINT `node_vector_constraint` FOR (n:`Movie`) REQUIRE (n.`embedding`) IS :: VECTOR<INTEGER32 NOT NULL>(42)"         |
```

| 39 | "part_of_tags"          | "RELATIONSHIP_PROPERTY_TYPE"       | "RELATIONSHIP" | ["PART_OF"]    | ["tags"]                     | NULL          | "independent"  | NULL                    | "STRING | LIST<STRING NOT NULL>" | NULL              | "CREATE CONSTRAINT `part_of_tags` FOR ()-[r:`PART_OF`]-() REQUIRE (r.`tags`) IS :: STRING | LIST<STRING NOT NULL>" |
| 24 | "prequels"              | "RELATIONSHIP_PROPERTY_UNIQUENESS" | "RELATIONSHIP" | ["PREQUEL_OF"] | ["order", "author"]          | NULL          | "undesignated" | "prequels"              | NULL                             | {indexConfig: {}} | "CREATE CONSTRAINT `prequels` FOR ()-[r:`PREQUEL_OF`]-() REQUIRE (r.`order`, r.`author`) IS UNIQUE"                |
| 30 | "rel_exist_param"       | "RELATIONSHIP_PROPERTY_EXISTENCE"  | "RELATIONSHIP" | ["WROTE"]      | ["published"]                | NULL          | "independent"  | NULL                    | NULL                             | NULL              | "CREATE CONSTRAINT `rel_exist_param` FOR ()-[r:`WROTE`]-() REQUIRE (r.`published`) IS NOT NULL"                    |

> | 3  | "rel_vector_constraint" | "RELATIONSHIP_PROPERTY_TYPE"       | "RELATIONSHIP" | ["CONTAINS"]   | ["embedding"]                | NULL          | ""             | NULL                    | "VECTOR<FLOAT32 NOT NULL>(1536)" | NULL              | "CREATE CONSTRAINT `rel_vector_constraint` FOR ()-[r:`CONTAINS`]-() REQUIRE (r.`embedding`) IS :: VECTOR<FLOAT32 NOT NULL>(1536)" |

Author: what's the value for "classification" here?

Reviewer: As with the node one, that depends on how they get created but assuming they are created using the `CREATE CONSTRAINT` command, it would be `"independent"`
```suggestion
| 3  | "rel_vector_constraint" | "RELATIONSHIP_PROPERTY_TYPE"       | "RELATIONSHIP" | ["CONTAINS"]   | ["embedding"]                | NULL          | "independent"             | NULL                    | "VECTOR<FLOAT32 NOT NULL>(1536)" | NULL              | "CREATE CONSTRAINT `rel_vector_constraint` FOR ()-[r:`CONTAINS`]-() REQUIRE (r.`embedding`) IS :: VECTOR<FLOAT32 NOT NULL>(1536)" |
```

> .Result
> [source, queryresult]
> ----
> +----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
> | id | name                     | type                               | entityType     | labelsOrTypes  | properties                   | ownedIndex              | propertyType                     | options           | createStatement                                                                                                                   |
> +----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
> | 23 | "actor_fullname"         | "NODE_KEY"                         | "NODE"         | ["Actor"]      | ["firstname", "surname"]     | "actor_fullname"        | NULL                             | {indexConfig: {}} | "CREATE CONSTRAINT `actor_fullname` FOR (n:`Actor`) REQUIRE (n.`firstname`, n.`surname`) IS KEY"                                  |
> | 12 | "author_name"            | "NODE_PROPERTY_EXISTENCE"          | "NODE"         | ["Author"]     | ["name"]                     | NULL                    | NULL                             | NULL              | "CREATE CONSTRAINT `author_name` FOR (n:`Author`) REQUIRE (n.`name`) IS NOT NULL"                                                 |
> | 5  | "book_isbn"              | "NODE_PROPERTY_UNIQUENESS"         | "NODE"         | ["Book"]       | ["isbn"]                     | "book_isbn"             | NULL                             | {indexConfig: {}} | "CREATE CONSTRAINT `book_isbn` FOR (n:`Book`) REQUIRE (n.`isbn`) IS UNIQUE"                                                       |
> | 9  | "book_title_year"        | "NODE_PROPERTY_UNIQUENESS"         | "NODE"         | ["Book"]       | ["title", "publicationYear"] | "book_title_year"       | NULL                             | {indexConfig: {}} | "CREATE CONSTRAINT `book_title_year` FOR (n:`Book`) REQUIRE (n.`title`, n.`publicationYear`) IS UNIQUE"                           |
> | 19 | "director_imdbId"        | "NODE_KEY"                         | "NODE"         | ["Director"]   | ["imdbId"]                   | "director_imdbId"       | NULL                             | {indexConfig: {}} | "CREATE CONSTRAINT `director_imdbId` FOR (n:`Director`) REQUIRE (n.`imdbId`) IS KEY"                                              |
> | 25 | "knows_since_how"        | "RELATIONSHIP_KEY"                 | "RELATIONSHIP" | ["KNOWS"]      | ["since", "how"]             | "knows_since_how"       | NULL                             | {indexConfig: {}} | "CREATE CONSTRAINT `knows_since_how` FOR ()-[r:`KNOWS`]-() REQUIRE (r.`since`, r.`how`) IS KEY"                                   |
> | 16 | "movie_tagline"          | "NODE_PROPERTY_TYPE"               | "NODE"         | ["Movie"]      | ["tagline"]                  | NULL                    | "STRING | LIST<STRING NOT NULL>" | NULL              | "CREATE CONSTRAINT `movie_tagline` FOR (n:`Movie`) REQUIRE (n.`tagline`) IS :: STRING | LIST<STRING NOT NULL>"                    |
> | 14 | "movie_title"            | "NODE_PROPERTY_TYPE"               | "NODE"         | ["Movie"]      | ["title"]                    | NULL                    | "STRING"                         | NULL              | "CREATE CONSTRAINT `movie_title` FOR (n:`Movie`) REQUIRE (n.`title`) IS :: STRING"                                                |
> | 27 | "node_uniqueness_param"  | "NODE_PROPERTY_UNIQUENESS"         | "NODE"         | ["Book"]       | ["prop1"]                    | "node_uniqueness_param" | NULL                             | {indexConfig: {}} | "CREATE CONSTRAINT `node_uniqueness_param` FOR (n:`Book`) REQUIRE (n.`prop1`) IS UNIQUE"                                          |
> | 2  | "node_vector_constraint" | "NODE_PROPERTY_TYPE"               | "NODE"         | ["Movie"]      | ["embedding"]                | NULL                    | "VECTOR<INTEGER32 NOT NULL>(42)" | NULL              | "CREATE CONSTRAINT `node_vector_constraint` FOR (n:`Movie`) REQUIRE (n.`embedding`) IS :: VECTOR<INTEGER32 NOT NULL>(42)"         |
> | 21 | "ownershipId"            | "RELATIONSHIP_KEY"                 | "RELATIONSHIP" | ["OWNS"]       | ["ownershipId"]              | "ownershipId"           | NULL                             | {indexConfig: {}} | "CREATE CONSTRAINT `ownershipId` FOR ()-[r:`OWNS`]-() REQUIRE (r.`ownershipId`) IS KEY"                                           |
> | 15 | "part_of"                | "RELATIONSHIP_PROPERTY_TYPE"       | "RELATIONSHIP" | ["PART_OF"]    | ["order"]                    | NULL                    | "INTEGER"                        | NULL              | "CREATE CONSTRAINT `part_of` FOR ()-[r:`PART_OF`]-() REQUIRE (r.`order`) IS :: INTEGER"                                           |
> | 17 | "part_of_tags"           | "RELATIONSHIP_PROPERTY_TYPE"       | "RELATIONSHIP" | ["PART_OF"]    | ["tags"]                     | NULL                    | "STRING | LIST<STRING NOT NULL>" | NULL              | "CREATE CONSTRAINT `part_of_tags` FOR ()-[r:`PART_OF`]-() REQUIRE (r.`tags`) IS :: STRING | LIST<STRING NOT NULL>"                |
> | 11 | "prequels"               | "RELATIONSHIP_PROPERTY_UNIQUENESS" | "RELATIONSHIP" | ["PREQUEL_OF"] | ["order", "author"]          | "prequels"              | NULL                             | {indexConfig: {}} | "CREATE CONSTRAINT `prequels` FOR ()-[r:`PREQUEL_OF`]-() REQUIRE (r.`order`, r.`author`) IS UNIQUE"                               |
> | 28 | "rel_exist_param"        | "RELATIONSHIP_PROPERTY_EXISTENCE"  | "RELATIONSHIP" | ["WROTE"]      | ["published"]                | NULL                    | NULL                             | NULL              | "CREATE CONSTRAINT `rel_exist_param` FOR ()-[r:`WROTE`]-() REQUIRE (r.`published`) IS NOT NULL"                                   |
> | 3  | "rel_vector_constraint"  | "RELATIONSHIP_PROPERTY_TYPE"       | "RELATIONSHIP" | ["CONTAINS"]   | ["embedding"]                | NULL                    | "VECTOR<FLOAT32 NOT NULL>(1536)" | NULL              | "CREATE CONSTRAINT `rel_vector_constraint` FOR ()-[r:`CONTAINS`]-() REQUIRE (r.`embedding`) IS :: VECTOR<FLOAT32 NOT NULL>(1536)" |
> | 7  | "sequels"                | "RELATIONSHIP_PROPERTY_UNIQUENESS" | "RELATIONSHIP" | ["SEQUEL_OF"]  | ["order"]                    | "sequels"               | NULL                             | {indexConfig: {}} | "CREATE CONSTRAINT `sequels` FOR ()-[r:`SEQUEL_OF`]-() REQUIRE (r.`order`) IS UNIQUE"                                             |
> | 13 | "wrote_year"             | "RELATIONSHIP_PROPERTY_EXISTENCE"  | "RELATIONSHIP" | ["WROTE"]      | ["year"]                     | NULL                    | NULL                             | NULL              | "CREATE CONSTRAINT `wrote_year` FOR ()-[r:`WROTE`]-() REQUIRE (r.`year`) IS NOT NULL"                                             |
> +----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
> ----

Author: ```suggestion
.Result
[source, queryresult]
----
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| id | name                    | type                               | entityType     | labelsOrTypes  | properties                   | enforcedLabel | classification | ownedIndex              | propertyType                     | options           | createStatement                                                                                                    |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| 28 | "actor_fullname"        | "NODE_KEY"                         | "NODE"         | ["Actor"]      | ["firstname", "surname"]     | NULL          | "undesignated" | "actor_fullname"        | NULL                             | {indexConfig: {}} | "CREATE CONSTRAINT `actor_fullname` FOR (n:`Actor`) REQUIRE (n.`firstname`, n.`surname`) IS KEY"                   |
| 27 | "author_name"           | "NODE_PROPERTY_EXISTENCE"          | "NODE"         | ["Author"]     | ["name"]                     | NULL          | "independent"  | NULL                    | NULL                             | NULL              | "CREATE CONSTRAINT `author_name` FOR (n:`Author`) REQUIRE (n.`name`) IS NOT NULL"                                  |
| 34 | "book_isbn"             | "NODE_PROPERTY_UNIQUENESS"         | "NODE"         | ["Book"]       | ["isbn"]                     | NULL          | "undesignated" | "book_isbn"             | NULL                             | {indexConfig: {}} | "CREATE CONSTRAINT `book_isbn` FOR (n:`Book`) REQUIRE (n.`isbn`) IS UNIQUE"                                        |
| 23 | "book_title_year"       | "NODE_PROPERTY_UNIQUENESS"         | "NODE"         | ["Book"]       | ["title", "publicationYear"] | NULL          | "undesignated" | "book_title_year"       | NULL                             | {indexConfig: {}} | "CREATE CONSTRAINT `book_title_year` FOR (n:`Book`) REQUIRE (n.`title`, n.`publicationYear`) IS UNIQUE"            |
| 36 | "director_imdbId"       | "NODE_KEY"                         | "NODE"         | ["Director"]   | ["imdbId"]                   | NULL          | "undesignated" | "director_imdbId"       | NULL                             | {indexConfig: {}} | "CREATE CONSTRAINT `director_imdbId` FOR (n:`Director`) REQUIRE (n.`imdbId`) IS KEY"                               |
| 26 | "knows_since_how"       | "RELATIONSHIP_KEY"                 | "RELATIONSHIP" | ["KNOWS"]      | ["since", "how"]             | NULL          | "undesignated" | "knows_since_how"       | NULL                             | {indexConfig: {}} | "CREATE CONSTRAINT `knows_since_how` FOR ()-[r:`KNOWS`]-() REQUIRE (r.`since`, r.`how`) IS KEY"                    |
| 12 | "movie_tagline"         | "NODE_PROPERTY_TYPE"               | "NODE"         | ["Movie"]      | ["tagline"]                  | NULL          | "independent"  | NULL                    | "STRING | LIST<STRING NOT NULL>" | NULL              | "CREATE CONSTRAINT `movie_tagline` FOR (n:`Movie`) REQUIRE (n.`tagline`) IS :: STRING | LIST<STRING NOT NULL>"     |
| 33 | "movie_title"           | "NODE_PROPERTY_TYPE"               | "NODE"         | ["Movie"]      | ["title"]                    | NULL          | "independent"  | NULL                    | "STRING"                         | NULL              | "CREATE CONSTRAINT `movie_title` FOR (n:`Movie`) REQUIRE (n.`title`) IS :: STRING"                                 |
| 17 | "node_uniqueness_param" | "NODE_PROPERTY_UNIQUENESS"         | "NODE"         | ["Book"]       | ["prop1"]                    | NULL          | "undesignated" | "node_uniqueness_param" | NULL                             | {indexConfig: {}} | "CREATE CONSTRAINT `node_uniqueness_param` FOR (n:`Book`) REQUIRE (n.`prop1`) IS UNIQUE"                           |
| 38 | "ownershipId"           | "RELATIONSHIP_KEY"                 | "RELATIONSHIP" | ["OWNS"]       | ["ownershipId"]              | NULL          | "undesignated" | "ownershipId"           | NULL                             | {indexConfig: {}} | "CREATE CONSTRAINT `ownershipId` FOR ()-[r:`OWNS`]-() REQUIRE (r.`ownershipId`) IS KEY"                            |
| 41 | "part_of"               | "RELATIONSHIP_PROPERTY_TYPE"       | "RELATIONSHIP" | ["PART_OF"]    | ["order"]                    | NULL          | "independent"  | NULL                    | "INTEGER"                        | NULL              | "CREATE CONSTRAINT `part_of` FOR ()-[r:`PART_OF`]-() REQUIRE (r.`order`) IS :: INTEGER"                            |
| 39 | "part_of_tags"          | "RELATIONSHIP_PROPERTY_TYPE"       | "RELATIONSHIP" | ["PART_OF"]    | ["tags"]                     | NULL          | "independent"  | NULL                    | "STRING | LIST<STRING NOT NULL>" | NULL              | "CREATE CONSTRAINT `part_of_tags` FOR ()-[r:`PART_OF`]-() REQUIRE (r.`tags`) IS :: STRING | LIST<STRING NOT NULL>" |
| 24 | "prequels"              | "RELATIONSHIP_PROPERTY_UNIQUENESS" | "RELATIONSHIP" | ["PREQUEL_OF"] | ["order", "author"]          | NULL          | "undesignated" | "prequels"              | NULL                             | {indexConfig: {}} | "CREATE CONSTRAINT `prequels` FOR ()-[r:`PREQUEL_OF`]-() REQUIRE (r.`order`, r.`author`) IS UNIQUE"                |
| 30 | "rel_exist_param"       | "RELATIONSHIP_PROPERTY_EXISTENCE"  | "RELATIONSHIP" | ["WROTE"]      | ["published"]                | NULL          | "independent"  | NULL                    | NULL                             | NULL              | "CREATE CONSTRAINT `rel_exist_param` FOR ()-[r:`WROTE`]-() REQUIRE (r.`published`) IS NOT NULL"                    |
| 19 | "sequels"               | "RELATIONSHIP_PROPERTY_UNIQUENESS" | "RELATIONSHIP" | ["SEQUEL_OF"]  | ["order"]                    | NULL          | "undesignated" | "sequels"               | NULL                             | {indexConfig: {}} | "CREATE CONSTRAINT `sequels` FOR ()-[r:`SEQUEL_OF`]-() REQUIRE (r.`order`) IS UNIQUE"                              |
| 14 | "wrote_year"            | "RELATIONSHIP_PROPERTY_EXISTENCE"  | "RELATIONSHIP" | ["WROTE"]      | ["year"]                     | NULL          | "independent"  | NULL                    | NULL                             | NULL              | "CREATE CONSTRAINT `wrote_year` FOR ()-[r:`WROTE`]-() REQUIRE (r.`year`) IS NOT NULL"                              |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
----
```

| 26 | "knows_since_how" | "RELATIONSHIP_KEY" | "RELATIONSHIP" | ["KNOWS"]     | ["since", "how"]         | NULL          | "knows_since_how" | NULL         |
| 38 | "ownershipId"     | "RELATIONSHIP_KEY" | "RELATIONSHIP" | ["OWNS"]      | ["ownershipId"]          | NULL          | "ownershipId"     | NULL         |
+------------------------------------------------------------------------------------------------------------------------------------------------------------+
----

For a full list of all the constraint types (and synonyms) available in this command see xref:schema/syntax.adoc#list-constraints[Syntax -> SHOW CONSTRAINTS].

======

.Filtering constraints using the `WHERE` clause
======

.List only constraints with a `RELATIONSHIP` `entityType`
[source, cypher, test-exclude-cols=id]
----
SHOW CONSTRAINTS
WHERE entityType = 'RELATIONSHIP'
----

.Result
[source, queryresult]
----
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+

> | id | name                    | type                               | entityType     | labelsOrTypes  | properties          | ownedIndex        | propertyType                     |

Reviewer: missing the `enforcedLabel` column

> .Result
> [source, queryresult]
> ----
> +----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
> | id | name                    | type                               | entityType     | labelsOrTypes  | properties          | ownedIndex        | propertyType                     |
> +----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
> | 25 | "knows_since_how"       | "RELATIONSHIP_KEY"                 | "RELATIONSHIP" | ["KNOWS"]      | ["since", "how"]    | "knows_since_how" | NULL                             |
> | 21 | "ownershipId"           | "RELATIONSHIP_KEY"                 | "RELATIONSHIP" | ["OWNS"]       | ["ownershipId"]     | "ownershipId"     | NULL                             |
> | 15 | "part_of"               | "RELATIONSHIP_PROPERTY_TYPE"       | "RELATIONSHIP" | ["PART_OF"]    | ["order"]           | NULL              | "INTEGER"                        |
> | 17 | "part_of_tags"          | "RELATIONSHIP_PROPERTY_TYPE"       | "RELATIONSHIP" | ["PART_OF"]    | ["tags"]            | NULL              | "STRING | LIST<STRING NOT NULL>" |
> | 11 | "prequels"              | "RELATIONSHIP_PROPERTY_UNIQUENESS" | "RELATIONSHIP" | ["PREQUEL_OF"] | ["order", "author"] | "prequels"        | NULL                             |
> | 28 | "rel_exist_param"       | "RELATIONSHIP_PROPERTY_EXISTENCE"  | "RELATIONSHIP" | ["WROTE"]      | ["published"]       | NULL              | NULL                             |
> | 3  | "rel_vector_constraint" | "RELATIONSHIP_PROPERTY_TYPE"       | "RELATIONSHIP" | ["CONTAINS"]   | ["embedding"]       | NULL              | "VECTOR<FLOAT32 NOT NULL>(1536)" |
> | 7  | "sequels"               | "RELATIONSHIP_PROPERTY_UNIQUENESS" | "RELATIONSHIP" | ["SEQUEL_OF"]  | ["order"]           | "sequels"         | NULL                             |
> | 13 | "wrote_year"            | "RELATIONSHIP_PROPERTY_EXISTENCE"  | "RELATIONSHIP" | ["WROTE"]      | ["year"]            | NULL              | NULL                             |
> +----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
> ----

Author: ```suggestion
.Result
[source, queryresult]
----
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| id | name              | type                               | entityType     | labelsOrTypes  | properties          | enforcedLabel | ownedIndex        | propertyType                     |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| 26 | "knows_since_how" | "RELATIONSHIP_KEY"                 | "RELATIONSHIP" | ["KNOWS"]      | ["since", "how"]    | NULL          | "knows_since_how" | NULL                             |
| 38 | "ownershipId"     | "RELATIONSHIP_KEY"                 | "RELATIONSHIP" | ["OWNS"]       | ["ownershipId"]     | NULL          | "ownershipId"     | NULL                             |
| 41 | "part_of"         | "RELATIONSHIP_PROPERTY_TYPE"       | "RELATIONSHIP" | ["PART_OF"]    | ["order"]           | NULL          | NULL              | "INTEGER"                        |
| 39 | "part_of_tags"    | "RELATIONSHIP_PROPERTY_TYPE"       | "RELATIONSHIP" | ["PART_OF"]    | ["tags"]            | NULL          | NULL              | "STRING | LIST<STRING NOT NULL>" |
| 24 | "prequels"        | "RELATIONSHIP_PROPERTY_UNIQUENESS" | "RELATIONSHIP" | ["PREQUEL_OF"] | ["order", "author"] | NULL          | "prequels"        | NULL                             |
| 30 | "rel_exist_param" | "RELATIONSHIP_PROPERTY_EXISTENCE"  | "RELATIONSHIP" | ["WROTE"]      | ["published"]       | NULL          | NULL              | NULL                             |
| 19 | "sequels"         | "RELATIONSHIP_PROPERTY_UNIQUENESS" | "RELATIONSHIP" | ["SEQUEL_OF"]  | ["order"]           | NULL          | "sequels"         | NULL                             |
| 14 | "wrote_year"      | "RELATIONSHIP_PROPERTY_EXISTENCE"  | "RELATIONSHIP" | ["WROTE"]      | ["year"]            | NULL          | NULL              | NULL                             |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
----
```

| LIST<STRING>

| enforcedLabel
| The implied labels required by a node label existence constraint or the source node and target node labels required by a relationship source or target label constraint, or `null` for other constraints which do not constrain labels. label:default-output[] label:cypher[Cypher 25 only] label:new[Introduced in Neo4j 2025.xx]
| STRING

| ownedIndex
| The name of the index associated with the constraint or `null`, in case no index is associated with it. label:default-output[]
| STRING

| propertyType
| The property type the property is restricted to for property type constraints, or `null` for the other constraints.
label:default-output[]
| STRING

| classification
| The status of a constraint in a graph type (`dependent`, `undesignated`, or `independent`). label:cypher[Cypher 25 only] label:new[Introduced in Neo4j 2025.xx]
| STRING

| options
| The options passed to `CREATE` command, for the index associated to the constraint, or `null` if no index is associated with the constraint.
| MAP

| createStatement

> | Statement used to create the constraint, or `null` for constraints with `dependent` classification or if the constraint cannot be created through `CREATE CONSTRAINT` (for example in case of version incompatibilities).

Reviewer: The reason for it being null is that it can't be created through create constraints, then both the version incompatibilities and dependent classifications are examples of why they can't be created.
```suggestion
| Statement used to create the constraint, or `null` if the constraint cannot be created through `CREATE CONSTRAINT` (for example in case of version incompatibilities or if for constraints with `dependent` classification).
```

| 36 | "director_imdbId"       | "NODE_KEY"                         | "NODE"         | ["Director"]   | ["imdbId"]                   | NULL          | "director_imdbId"       | NULL                             |
| 26 | "knows_since_how"       | "RELATIONSHIP_KEY"                 | "RELATIONSHIP" | ["KNOWS"]      | ["since", "how"]             | NULL          | "knows_since_how"       | NULL                             |
| 12 | "movie_tagline"         | "NODE_PROPERTY_TYPE"               | "NODE"         | ["Movie"]      | ["tagline"]                  | NULL          | NULL                    | "STRING | LIST<STRING NOT NULL>" |
| 33 | "movie_title"           | "NODE_PROPERTY_TYPE"               | "NODE"         | ["Movie"]      | ["title"]                    | NULL          | NULL                    | "STRING"                         |
| 17 | "node_uniqueness_param" | "NODE_PROPERTY_UNIQUENESS"         | "NODE"         | ["Book"]       | ["prop1"]                    | NULL          | "node_uniqueness_param" | NULL                             |

> | 2  | "node_vector_constraint" | "NODE_PROPERTY_TYPE"              | "NODE"         | ["Movie"]      | ["embedding"]                | NULL          | NULL                    | "VECTOR<INTEGER32 NOT NULL>(42)" |

Reviewer: what about this table?

| 14 | "wrote_year"            | "RELATIONSHIP_PROPERTY_EXISTENCE"  | "RELATIONSHIP" | ["WROTE"]      | ["year"]                     | NULL          | NULL                    | NULL                             |
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
----

======

.List all constraints with full details
======

To return the full details of the constraints on a database, use `SHOW CONSTRAINTS YIELD *`

.List all constraints with `YIELD *`
// tag::schema_constraints_show_full[]
[source, cypher, test-exclude-cols=id]
----
SHOW CONSTRAINTS YIELD *
----
// end::schema_constraints_show_full[]

.Result
[source, queryresult]
----
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+

> | id | name                    | type                               | entityType     | labelsOrTypes  | properties                   | enforcedLabel | classification | ownedIndex              | propertyType                     | options           | createStatement                                                                                                                    |

Reviewer: the header line wasn't aligned (I think this is correct 🤔)
```suggestion
| id | name                     | type                               | entityType     | labelsOrTypes  | properties                   | enforcedLabel | classification | ownedIndex              | propertyType                     | options           | createStatement                                                                                                                   |
```

|===
| Column | Description | Type

| id
| The id of the constraint. label:default-output[]
| INTEGER

| name
| Name of the constraint (explicitly set by the user or automatically assigned). label:default-output[]
| STRING

| type
| The ConstraintType of this constraint (`NODE_PROPERTY_UNIQUENESS`, `RELATIONSHIP_PROPERTY_UNIQUENESS`, `NODE_PROPERTY_EXISTENCE`, `RELATIONSHIP_PROPERTY_EXISTENCE`, `NODE_PROPERTY_TYPE`, `RELATIONSHIP_PROPERTY_TYPE`, `NODE_LABEL_EXISTENCE`, `RELATIONSHIP_SOURCE_LABEL`, `RELATIONSHIP_TARGET_LABEL`, `NODE_KEY`, or `RELATIONSHIP_KEY`). label:default-output[]

| STRING

| entityType
| Type of entities this constraint represents (`NODE` or `RELATIONSHIP`). label:default-output[]
| STRING

| labelsOrTypes
| The labels or relationship types of this constraint.
The list returned will only include a single value (the name of the constrained node label or relationship type). label:default-output[]
| LIST<STRING>

| properties
| The properties of this constraint, or `null` for node label existence, relationship source label, and relationship target constraints which do not constrain properties. label:default-output[]
| LIST<STRING>

| enforcedLabel

> | The implied labels required by a node label existence constraint or the source node and target node labels required by a relationship source or target label constraint, or `null` for other constraints which do not constrain labels. label:default-output[] label:cypher[Cypher 25 only] label:new[Introduced in Neo4j 2025.xx]

Reviewer: ```suggestion
| The implied labels required by a node label existence constraint or the source node and target node labels required by a relationship source or target label constraint, or `null` for other constraints which do not constrain labels. label:default-output[] label:cypher[Cypher 25 only] label:new[Introduced in Neo4j 2026.02]
```

| STRING

| ownedIndex
| The name of the index associated with the constraint or `null`, in case no index is associated with it. label:default-output[]
| STRING

| propertyType
| The property type the property is restricted to for property type constraints, or `null` for the other constraints.
label:default-output[]
| STRING

| classification

> | The status of a constraint in a graph type (`dependent`, `undesignated`, or `independent`). label:cypher[Cypher 25 only] label:new[Introduced in Neo4j 2025.xx]

Reviewer: ```suggestion
| The status of a constraint in a graph type (`dependent`, `undesignated`, or `independent`). label:cypher[Cypher 25 only] label:new[Introduced in Neo4j 2026.02]
```

| STRING

| options
| The options passed to `CREATE` command, for the index associated to the constraint, or `null` if no index is associated with the constraint.
| MAP

| createStatement

> | Statement used to create the constraint, or `null` if the constraint cannot be created through `CREATE CONSTRAINT` (for example in case of version incompatibilities or if for constraints with `dependent` classification).

Reviewer: That `if` looks out of place XD
```suggestion
| Statement used to create the constraint, or `null` if the constraint cannot be created through `CREATE CONSTRAINT` (for example in case of version incompatibilities or for constraints with `dependent` classification).
```

.Result
[source, queryresult]
----
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+

> | id | name                    | type                               | entityType     | labelsOrTypes  | properties                   | enforcedLabel | ownedIndex              | propertyType                     |

Reviewer: for all three of these tables, shouldn't they also include the vector constraints? We have a sentence at the top of the file that says
```
The constraints listed on this page are created on the page xref:schema/constraints/create-constraints.adoc[].
```
so I would think the new vector ones should be listed as well as the graph type related updates 🤔 

 (might need to be added to the setup at the start of the file to actually exist, `The below constraints are created on schema/constraints/create-constraints.adoc`)

Reviewer: the 4th table (that explicitly returns a subset of columns and therefor didn't need graph type updates) still have the vector ones as well so as it stands right now it's not even consistent 🤷

Author: did a diff check and updated. some table cells i'm not too sure about, will add comments for those

| 12 | "movie_tagline"         | "NODE_PROPERTY_TYPE"               | "NODE"         | ["Movie"]      | ["tagline"]                  | NULL          | "independent"  | NULL                    | "STRING | LIST<STRING NOT NULL>" | NULL              | "CREATE CONSTRAINT `movie_tagline` FOR (n:`Movie`) REQUIRE (n.`tagline`) IS :: STRING | LIST<STRING NOT NULL>"     |
| 33 | "movie_title"           | "NODE_PROPERTY_TYPE"               | "NODE"         | ["Movie"]      | ["title"]                    | NULL          | "independent"  | NULL                    | "STRING"                         | NULL              | "CREATE CONSTRAINT `movie_title` FOR (n:`Movie`) REQUIRE (n.`title`) IS :: STRING"                                 |
| 17 | "node_uniqueness_param" | "NODE_PROPERTY_UNIQUENESS"         | "NODE"         | ["Book"]       | ["prop1"]                    | NULL          | "undesignated" | "node_uniqueness_param" | NULL                             | {indexConfig: {}} | "CREATE CONSTRAINT `node_uniqueness_param` FOR (n:`Book`) REQUIRE (n.`prop1`) IS UNIQUE"                           |

> | 2  | "node_vector_constraint" | "NODE_PROPERTY_TYPE"               | "NODE"         | ["Movie"]      | ["embedding"]                | NULL         | "independent"             | NULL                    | "VECTOR<INTEGER32 NOT NULL>(42)" | NULL              | "CREATE CONSTRAINT `node_vector_constraint` FOR (n:`Movie`) REQUIRE (n.`embedding`) IS :: VECTOR<INTEGER32 NOT NULL>(42)"         |

Reviewer: The vector rows are not aligned along the columns in this table, might need to be updated (might be that the others need to be extended to fit the vector ones as well, but at least some is on the vector lines to fix as well)
<img width="1035" height="786" alt="image" src="https://github.com/user-attachments/assets/83247ad0-399c-41b6-b8f1-a0dc6e2c7eca" />
<img width="1503" height="772" alt="image" src="https://github.com/user-attachments/assets/ee00de24-d952-450b-a606-6ff40b532b51" />
<img width="1035" height="786" alt="image" src="https://github.com/user-attachments/assets/c608d84e-526d-4c50-9171-151dfec6eee8" />

this is also likely to be an issue for the other result tables and not just this one

Reviewer: what I would start with would be to align all rows for the name (adding a space/dash to the ones that need it and not remove the one from the node vector one), then remove the spacing after `"independent"` on the two vector ones to align there, and then check if the ends need some update to align all the rows with the widest column (and checking that all column lines are then aligned for all rows)

Author: should be aligned now
